crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+                    _ => return false,
+                },
+                PyOperationType(operation) => {
+                    match Python::with_gil(|py| -> Option<Array2<Complex64>> {
+                        Self::get_op(py, &operation.operation)
+                    }) {
+                        Some(x) => x,
+                        _ => return false,
+                    }
+                }
+                _ => return false,
+            },
+        };
+
+        if first_qarg == second_qarg {
+            abs_diff_eq!(
+                second_mat.dot(&first_mat),
+                first_mat.dot(&second_mat),
+                epsilon = 1e-8
+            )
+        } else {
+            let extra_qarg2 = num_qubits - first_qarg.len();
+            let first_mat = if extra_qarg2 > 0 {
+                let id_op = Array2::<Complex64>::eye(usize::pow(2, extra_qarg2 as u32));
+                kron(&id_op, &first_mat)
+            } else {
+                first_mat
+            };
+            let op12 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, false);
+            let op21 = compose(&first_mat.view(), &second_mat.view(), &second_qarg, true);
+            abs_diff_eq!(op12, op21, epsilon = 1e-8)
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn commutation_precheck(
+        &self,
+        op1: &CircuitInstruction,
+        qargs1: &[usize],
+        cargs1: &[usize],
+        op2: &CircuitInstruction,
+        qargs2: &[usize],
+        cargs2: &[usize],
+        max_num_qubits: u32,
+    ) -> Option<bool> {
+        if op1.op().control_flow()
+            || op2.op().control_flow()
+            || op1.is_conditioned()
+            || op2.is_conditioned()
+        {
+            return Some(false);
+        }
+
+        // assuming the number of involved qubits to be small, this might be faster than set operations
+        if !qargs1.iter().any(|e| qargs2.contains(e)) && !cargs1.iter().any(|e| cargs2.contains(e))
+        {
+            return Some(true);
+        }
+
+        if qargs1.len() > max_num_qubits as usize || qargs2.len() > max_num_qubits as usize {
+            return Some(false);
+        }
+
+        if SUPPORTED_OP.contains(&op1.op().name()) && SUPPORTED_OP.contains(&op2.op().name()) {
+            return None;
+        }
+
+        if self.is_commutation_skipped(op1) || self.is_commutation_skipped(op2) {
+            return Some(false);
+        }
+
+        None
+    }
+
+    fn get_op(py: Python, gate: &PyObject) -> Option<Array2<Complex64>> {
+        Some(
+            QI_OPERATOR
+                .get_bound(py)
+                .call1((gate,))
+                .ok()?
+                .getattr(intern!(py, "data"))
+                .ok()?
+                .extract::<PyReadonlyArray2<Complex64>>()
+                .ok()?
+                .as_array()
+                .to_owned(),
+        )
+    }
+
+    fn is_commutation_skipped(&self, instr: &CircuitInstruction) -> bool {
```

Comment:
Does this method need to be a class method? It doesn't use `&self` here for anything.